sdk/core/src/pipeline.rs

```diff
@@ -0,0 +1,49 @@
+use crate::policies::{Policy, PolicyResult};
+use crate::{Context, Request, Response};
+use std::sync::Arc;
+
+/// Execution pipeline.
+///
+/// A pipeline follows a precise flow:
+///
+/// 1. Per call policies are executed. Per call policies can fail and bail out of the pipeline
+///    immediately.
+/// 2. Retry policy. It allows to reexecute the following policies.
+/// 3. Per retry policies. Per retry polices are always executed at least once but are reexecuted
+///    in case of retries.
+/// 4. Transport policy. Transtport policy is always the last policy and is the policy that
+///    actually constructs the `Response` to be passed up the pipeline.
+///
+/// A pipeline is immutable. In other words a policy can either succeed and call the following
+/// policy of fail and return to the calling policy. Arbitrary policy "skip" must be avoided (but
+/// cannot be enforced by code). All policies except Transport policy can assume there is another following policy (so
+/// self.pipeline[0] is always valid).
+#[derive(Debug, Clone)]
+pub struct Pipeline {
+    pipeline: Vec<Arc<dyn Policy>>,
+}
+
+impl Pipeline {
+    pub fn new(
```

Comment:
FWIW, the only thing that should be able to create a pipeline is the client library, e.g. `KeyClient` (Key Vault keys client). However, end users should be able to add their own per-call or per-retry policies. We currently don't allow them to remove built-in ones (ones added by azure_core or the client class).

We probably don't need that here, but should think about how it would work. The previous link should show how we can pass them in. We could either make the ClientOptions classes merge vectors and pass them in, or maybe add 2 parameters here like `user_per_call_policies` and `user_per_retry_policies`. The idea of a builder of these like most (all?) of our languages use might also help make it easier.